src/interface/swarm.cpp

```diff
@@ -906,12 +890,32 @@ int Swarm::CountParticlesToSend_() {
           max_indices_size = num_particles_to_send_h(blockIndex_h(n));
         }
       }
+      if (blockIndex_h(n) == no_block_) {
+        total_noblock_particles++;
+      }
     }
   }
   // Size-0 arrays not permitted but we don't want to short-circuit subsequent logic that
   // indicates completed communications
   max_indices_size = std::max<int>(1, max_indices_size);
+
   // Not a ragged-right array, just for convenience
+  if (total_noblock_particles > 0) {
+    auto noblock_indices =
+        ParArrayND<int>("Particles with no block", total_noblock_particles);
+    auto noblock_indices_h = noblock_indices.GetHostMirror();
+    int counter = 0;
+    for (int n = 0; n <= max_active_index_; n++) {
+      if (mask_h(n)) {
+        if (blockIndex_h(n) == no_block_) {
+          noblock_indices_h(counter) = n;
+          counter++;
+        }
+      }
+    }
+    noblock_indices.DeepCopy(noblock_indices_h);
+    ApplyBoundaries_(total_noblock_particles, noblock_indices);
```

Comment:
I had broken the `particle_bounds_outflow` test here because the addition of the `no_block_` value when getting a neighbor block index (which is what got outflow boundaries working in the first place) meant that a particle wouldn't be sent across meshblocks which was the only code path triggering `ApplyBoundaries`. Unfortunately this adds *another* deep copy to the particles, which I will go ahead and clean up once `par_scan` is in parthenon (see #674 for the issue I just opened)

Reply:
General question: Why are all these host functions and not on the device?

Reply:
It wasn't obvious to me earlier how to do this in parallel in device code, but the answer is basically `parallel_scan` so once #667 goes in I will go back and rewrite all these host functions